baseplate/lib/experiments/__init__.py

```diff
@@ -137,9 +138,10 @@ def is_valid_experiment(self, name: str) -> bool:
 
     def variant(
```

Comment:
I'm not sure if this is worth the extra syntax and maybe fighting with linters (it's been a pain in the past) but `@typing.overload` can be nice for calling out mutually exclusive variants of a function's signature to the type system.

```python
@overload
def variant(self, name: str, user: Optional[User] = None, bucketing_event_override: Optional[bool] = None) -> Optional[str]:
    ...

@overload
def variant(self, name: str, user: Optional[User] = None, bucketing_event_override: Optional[bool] = None) -> Optional[str]:
    ...

def variant(self, experiment_name=None, user=None, bucketing_event_override=None, name=None):
    r"""The actual code..."""
```

(in use here: https://github.com/reddit/baseplate.py/blob/develop/baseplate/__init__.py#L475-L524)

Reply:
The overloads need to be "Keyword Only" functions in order for them to work with mypy. I think it makes sense in this case, that the overloads should be keyword only.  For typing purposes, keywords matter and ordering doesn't matter as much.

Error I got from mpy.
```
baseplate/lib/experiments/__init__.py:158: error: Overloaded function implementation does not accept all possible arguments of signature 1
```

Reply:
Oh I missed the `**kwargs: str` in the original method signature. Is that what it's complaining about in the overloads? I think we want to keep that as that's how you pass arbitrary variables in, right?

Reply:
I thought it wasn't as necessary, because the base signature would handle kwargs and any arguments.  But maybe for correctness, it makes sense to have the kwargs in the narrowed signatures.

Reply:
>Is that what it's complaining about in the overloads?

It wasn't related to kwargs, it's related to the order of arguments and their names.  Because if I switch the `name` and `experiment_name` arguments back with `name` as the first argument, then it complains that the second signature is incorrect.